src/coreComponents/denseLinearAlgebra/denseLASolvers.hpp

```diff
@@ -0,0 +1,349 @@
+
+    real64 const detX2 = A[0][0] * (A[1][1] * b[2] - A[1][2] * b[1]) -
+                         A[0][1] * (A[1][0] * b[2] - A[1][2] * b[0]) +
+                         A[0][2] * (A[1][0] * b[1] - A[1][1] * b[0]);
+
+    x[0] = detX0 / detA;
+    x[1] = detX1 / detA;
+    x[2] = detX2 / detA;
+}
+
+/**
+ * @brief Solves a linear system where the matrix is upper triangular using back substitution.
+ * 
+ * This function solves the linear system `Ax = b`, where `A` is an upper triangular matrix, using 
+ * back substitution. The solution `x` is computed and stored in the provided output vector.
+ * 
+ * @tparam N The size of the square matrix `A`.
+ * @tparam MATRIX_TYPE The type of the matrix `A`.
+ * @tparam RHS_TYPE The type of the right-hand side vector `b`.
+ * @tparam SOL_TYPE The type of the solution vector `x`.
+ * @param[in] A The upper triangular matrix representing the coefficients of the system.
+ * @param[in,out] b The right-hand side vector. It is used to compute the solution, but the values are not preserved.
+ * @param[out] x The solution vector. The result of solving the system `Ax = b` using back substitution.
+ */
+template< std::ptrdiff_t N, 
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE >
+void solveUpperTriangularSystem( MATRIX_TYPE const & A, RHS_TYPE & b, SOL_TYPE && x )
+{
+  for (std::ptrdiff_t i = N - 1; i >= 0; --i)
+  {
+    real64 sum = b[i];
+    for (std::ptrdiff_t j = i + 1; j < N; ++j)
+    {
+      sum -= A[i][j] * x[j];
+    }
+    x[i] = sum / A[i][i];
+  }
+}
+
+/**
+ * @brief Solves a linear system using Gaussian elimination.
+ * 
+ * This function performs Gaussian elimination on the given matrix `A` and right-hand side vector `b`.
+ * It transforms the matrix `A` into an upper triangular matrix and then solves for the solution `x`
+ * using back substitution.
+ * 
+ * @tparam N The size of the square matrix `A`.
+ * @tparam MATRIX_TYPE The type of the matrix `A`.
+ * @tparam RHS_TYPE The type of the right-hand side vector `b`.
+ * @tparam SOL_TYPE The type of the solution vector `x`.
+ * @param[in,out] A The matrix to be transformed into an upper triangular matrix. Modified in place.
+ * @param[in,out] b The right-hand side vector. Modified in place to reflect the transformed system.
+ * @param[out] x The solution vector. The result of solving the system `Ax = b`.
+ */
+template< std::ptrdiff_t N, 
+          typename MATRIX_TYPE,
+          typename RHS_TYPE,
+          typename SOL_TYPE >
+GEOS_HOST_DEVICE 
+inline
+void solveGaussianElimination( MATRIX_TYPE & A, RHS_TYPE & b, SOL_TYPE && x )
+{
+  static_assert( N > 0, "N must be greater than 0." );
+  internal::checkSizes< N, N >( matrix );
+  internal::checkSizes< N >( b ); 
+  internal::checkSizes< N >( x );
+  
+
+  // Step 1: Transform  into an upper triangular matrix 
+  
+  // 1.a. Find the pivot
+  for (std::ptrdiff_t i = 0; i < N; ++i)
```

Comment:
Minor detail: how large do we expect `N` to be? For large enough (depends on the GPU, but let's say a conservative 8x8), the current threading strategy will lead to register spilling and suboptimal performance.

Reply:
This is very important @victorapm. Once we agree on the value for `N`, we should throw an error if larger systems are used.

Reply:
that is a good point. I was thinking of adding a `static_assert` to enforce N to be smaller than a certain number (10?). What to do for larger systems is up for discussion. We would have to come up with a mechanism to change the threading strategy depending on the size of the system. Maybe it's possible through `raja`?